contracts/test/v0.8/dev/UpkeepTranscoder3_0.test.ts

```diff
@@ -0,0 +1,458 @@
+      ]
+
+      it('transcodes V1 upkeeps to V3 properly', async () => {
+        upkeepsV1 = [
+          [
+            balance,
+            lastKeeper0,
+            executeGas,
+            2 ** 32,
+            target0,
+            amountSpent,
+            await admin0.getAddress(),
+          ],
+          [
+            balance,
+            lastKeeper1,
+            executeGas,
+            2 ** 32,
+            target1,
+            amountSpent,
+            await admin1.getAddress(),
+          ],
+        ]
+
+        const data = await transcoder.transcodeUpkeeps(
+          UpkeepFormat.V1,
+          UpkeepFormat.V3,
+          encodeUpkeepV1(idx, upkeepsV1, ['0xabcd', '0xffff']),
+        )
+        assert.equal(
+          encodeUpkeepV3(idx, upkeepsV3, ['0xabcd', '0xffff'], admins),
+          data,
+        )
+      })
+
+      it('transcodes V2 upkeeps to V3 properly', async () => {
+        upkeepsV2 = [
+          [
+            balance,
+            lastKeeper0,
+            amountSpent,
+            await admin0.getAddress(),
+            executeGas,
+            2 ** 32 - 1,
+            target0,
+            false,
+          ],
+          [
+            balance,
+            lastKeeper1,
+            amountSpent,
+            await admin1.getAddress(),
+            executeGas,
+            2 ** 32 - 1,
+            target1,
+            false,
+          ],
+        ]
+
+        const data = await transcoder.transcodeUpkeeps(
+          UpkeepFormat.V2,
+          UpkeepFormat.V3,
+          encodeUpkeepV2(idx, upkeepsV2, ['0xabcd', '0xffff']),
+        )
+        assert.equal(
+          encodeUpkeepV3(idx, upkeepsV3, ['0xabcd', '0xffff'], admins),
+          data,
+        )
+      })
+
+      it('migrates upkeeps from one registry to another', async () => {
+        const [id, randomBytes, registry12, registry20] =
+          await deployRegistries()
+
+        // set outgoing permission to registry 2_0 and incoming permission for registry 1_2
+        await registry12.setPeerRegistryMigrationPermission(
+          registry20.address,
+          1,
+        )
+        await registry20.setPeerRegistryMigrationPermission(
+          registry12.address,
+          2,
+        )
+
+        expect((await registry12.getUpkeep(id)).balance).to.equal(toWei('1000'))
+        expect((await registry12.getUpkeep(id)).checkData).to.equal(randomBytes)
+        expect((await registry12.getState()).state.numUpkeeps).to.equal(1)
+
+        await registry12
+          .connect(admin0)
+          .migrateUpkeeps([id], registry20.address)
+
+        expect((await registry12.getState()).state.numUpkeeps).to.equal(0)
+        expect((await registry20.getState()).state.numUpkeeps).to.equal(1)
+        expect((await registry12.getUpkeep(id)).balance).to.equal(0)
+        expect((await registry12.getUpkeep(id)).checkData).to.equal('0x')
+        expect((await registry20.getUpkeep(id)).balance).to.equal(toWei('1000'))
+        expect(
```

Comment:
can we also verify that the link token is transferred? (Link.balanceOf(registry20) == 1000 wei)

Reply:
done